xlators/storage/posix/src/posix-inode-fd-ops.c

```diff
@@ -5823,7 +5819,7 @@ posix_fill_readdir(fd_t *fd, struct posix_fd *pfd, off_t off, size_t size,
         count++;
     }
 
-    if ((!sys_readdir(pfd->dir, scratch) && (errno == 0))) {
+    if (!entry && (errno == 0)) {
```

Comment:
This is the one that should be a bit more head scratching.  The loop can come out on three conditions:

1.  The while condition (which we can as far as I can tell replace with while(1) since filled can never be > size (based on 2).
2.  We can't add the next entry due to filled + this_size > size (ie, overflowing)
3.  Hitting EOF.

EOF is the ONLY case where the readdir() inside the loop will return NULL (and errno still == 0).

So there is no point in performing another extra readdir() here.

I trust that the semantics of glusterfs here is that we need to set errno=ENOENT to indicate EOF, rather than merely return a count of 0?  Anyway, I don't think this particularly matters, this block indentation will execute once per full readdir() cycle through a folder, compared to some other things.

That said, all of these changes are *marginal* improvements at best, although, the dropped memsets can in the case of DT_UNKNOWN probably be significant.  But I've not encountered DT_UNKNOWN recently anyway it still needs to be covered.

Reply:
DT_UNKNOWN is the least used path (as well as the change that is correct, but relevant for BSD only, which is why I did not change it).